### PR TITLE
SCRIPTS/SQL: Storms of Fate

### PR DESCRIPTION
**Pull request by Zenny89**
Special thanks to @Hozu  and @TeoTwawki

### CODE REVIEW

**Comment by Hozu** [comment]
The var for Megaflare/Gigaflare is already handled btw, so that part of the comment can be removed. Excellent job !

**Comment by teschnei** [comment]
I don't know if I'd bother with Bahamut's script yet.  You didn't make him do the talking and it would be really annoying to do before the AIRewrite.  Also, gigaflare and megaflare do not have the same cast time

**Comment by Zenny89** [comment]
I used this video to get cast times and they were the same: https://www.youtube.com/watch?v=XRoX7Vq_ZAQ
He still should do the moves every 10% though, regardless of the talking part.

**Comment by teschnei** [comment]
Okay, video is better proof than the shitty wiki.  I'll give you the option of either waiting for the AIRewrite to do the talking or scripting it in onMobFight now, because I'd really like it to be completely correct first

**Comment by Zenny89** [comment]
Okay, I understand - I'll work on the talking portion of it. Unrelated, I left a comment in the zone.lua of Ru'Lude Gardens that I'd like someone to look at.

**Comment by Hozu** [comment]
I know there's talking handled for the animated weapons already (engage/death/party wipe) so I'm curious as to why it would be difficult for Mega/Gigaflare?

File: scripts/globals/mobskills/Gigaflare.lua [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    if (mobhp <= 10) then
+        result = 0;
+    end;
+
+    return result;
+end;
+
+function onMobWeaponSkill(target, mob, skill)
+    local gigaflareCount = mob:getLocalVar("GigaFlare") + 1; -- increments the variable for scripted use of Gigaflare
```
Comment (TeoTwawki) [inline scripts/globals/mobskills/Gigaflare.lua]: Does he need a var to keep tracking how many went off after the first? 
https://github.com/DarkstarProject/darkstar/pull/2001/files#diff-87a705ccc8d0a281b22f4da231bc1972R55

Comment (Hozu) [inline scripts/globals/mobskills/Gigaflare.lua]: For the first one, if you stun it, it needs to try to use it again, so yeah.

**Comment by Zenny89** [comment]
In this case, I don't think so because he only uses 1 GigaFlare

File: scripts/globals/mobskills/Gigaflare.lua (hunk 1 above)
Comment (TeoTwawki) [inline scripts/globals/mobskills/Gigaflare.lua]: still wouldn't need incremented again. would need to not increment to 1 form zero till it actually foes off, for that scenario.

>

Comment (TeoTwawki) [inline scripts/globals/mobskills/Gigaflare.lua]: dammit fingers, wasn't done typing. (see edit)

Comment (Hozu) [inline scripts/globals/mobskills/Gigaflare.lua]: Yeah both Megaflare and Gigaflare will increment the variable for each unscripted use, but since there's no more scripted uses of either move after the one scripted Gigaflare I don't think it particularly matters, does it?

**Comment by teschnei** [comment]
"I know there's talking handled for the animated weapons already (engage/death/party wipe) so I'm curious as to why it would be difficult for Mega/Gigaflare?" 
the talking isn't the issue, the issue is the timing (there is no trigger for it because it's just during charging time)

**Comment by TeoTwawki** [comment]
@Hozu  all we can do is temp have him just say it before useMobAbility() for now.

 (well there are other ways..but with the airewrite coming up its not worth it to make huge workarounds to get the timing retail perfect).

**Comment by teschnei** [comment]
also yeah - you are using two different vars for tracking -flares, one in bahamut.lua and one in the weaponskill script - the vars should only be incremented when he actually casts it

**Comment by teschnei** [comment]
i think onMobFight still runs while charging mobskills, so you can use localvars to fake it there

**Comment by teschnei** [comment]
but I will do my best to make it super easy in the AI rewrite

**Comment by Hozu** [comment]
Hm I think the wiki may be wrong for the text for Megaflare - watching a video https://www.youtube.com/watch?v=nJ3Om5PVhjo seems to have it say the first text only once, the second text for every other Megaflare, and the third text might be only upon killing someone. I'll keep watching this video/others for more information.

**Comment by Hozu** [comment]
Ok there are three taunts in the text IDs. They are:
-Children of Vana'diel, what do you think to accomplish by fighting for your lives?
You know your efforts are futile, yet still you resist...
-Your lives are a blasphemy against the mothercrystal!
-Contemplate on the good your death will bring as your mortal form is consumed by my brilliance!

The way they appear to work is the first one is called, then two seconds later the second one is called, and Megaflare begins to prepare immediately afterwards. In between those two texts, no mobskills are used, and seemingly regular melee attacks stop until Megaflare is used. The third one is called simply on preparation of Gigaflare. All three texts are certainly doable right now. I haven't seen a video where either move is used under 10% simply because it tends to die very quickly at that point.

From watching various videos, Touchdown has slightly more knockback than Trample (probably 5), and Gigaflare seems to be radial AoE (https://youtu.be/fMpmrgKgBKQ?t=143).

**Comment by Zenny89** [comment]
So I should remove the increment for the flares in bahamut.lua and leave them only in the -flare.lua's?

**Comment by Hozu** [comment]
Oh I see what he's saying, yeah, remove the setting of the Megaflare/Gigaflare local var setting in bahamut.lua, otherwise it will increment twice for every scripted use.

**Comment by TeoTwawki** [comment]
Normally I'd say the other way but Bahamut here is a bit special because of that stun thing Hozu pointed out.

**Comment by teschnei** [comment]
yes, it needs to stay in the weaponskill script

**Comment by Hozu** [comment]
@Zenny89 I've been working on what you have to get it working right with the text and I nearly have it complete now. I'll show you the stuff when I can.

**Comment by Deadwing888** [comment]
I have screen capture footage of me screwing around with ouryu/bv2 if you all need a piece of information missing from wiki/youtube videos I can check mine.

**Comment by Hozu** [comment]
@Deadwing888 two questions - is Gigaflare actually AoE? I saw a video where someone who was behind Bahamut (but kind of off to the side) got hit by it. I'm wondering if it's AoE centered on the target.

Also, when it casts a -ga spell, it seemed to have a really large AoE radius. Have you noticed that as well?

**Comment by Zenny89** [comment]
Updated with Hozu's changes.

**Comment by Hozu** [comment]
ahhh! It wasn't entirely complete yet because I didn't handle the noturn behaviour thing. Now everyone can see I use tabs until final cleanup.

**Comment by Zenny89** [comment]
Yeah, I just noticed and am doing Tab -> Space :P

**Comment by Hozu** [comment]
@Zenny89 check IRC I sent you the noturn thing. With that it's complete.

**Comment by Zenny89** [comment]
Updated - hopefully good to go!

File: scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua [hunk 2]
```diff
@@ -1,15 +1,97 @@
...
+                    end;
+                    mob:useMobAbility(1295);
+                end;    
+            end;    
+        elseif (MegaFlareQueue == 0 and mobHPP < 10 and GigaFlare < 1 and mob:checkDistance(target) <= 15) then  -- All of the scripted Megaflares are to happen before Gigaflare.
+            if (tauntThreeShown == 0) then
+                target:showText(mob,BAHAMUT_TAUNT + 2);
+                mob:setLocalVar("tauntThreeShown", 1); -- again, taunt won't show again until the move is successfully used.
+            end;
+            -- disable noturn here
```
Comment (Hozu) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: Oops I forgot to put in the noturn thing in here too. Copy the thing at line 80 and replace the comment.

File: sql/mob_skill.sql [hunk 3]
```diff
@@ -179,8 +179,8 @@ INSERT INTO `mob_skill` VALUES (48,178,48,'Photosynthesis',0,7.0,2000,1500,1,0,0
 INSERT INTO `mob_skill` VALUES (49,178,49,'Leaf_Dagger',0,7.0,2000,1500,4,0,0,0);
 INSERT INTO `mob_skill` VALUES (44,178,44,'Head_Butt',0,7.0,1500,1500,4,0,0,0); -- occasional knockback..
 INSERT INTO `mob_skill` VALUES (50,178,50,'Scream',1,15.0,2000,1500,4,0,0,0);
+INSERT INTO `mob_skill` VALUES (1954,496,1603,'Petal_Pirouette',1,15.0,2000,1500,4,0,0,0); -- Only used by Chaneque
```
Comment (Hozu) [inline sql/mob_skill.sql]: er I don't think you did that? Might have snuck in from when I gave you the updated knockback values.

File: scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua [hunk 4]
```diff
@@ -1,15 +1,97 @@
...
+    local MegaFlareTrigger = mob:getLocalVar("MegaFlareTrigger");
+    local MegaFlareUses = mob:getLocalVar("MegaFlareUses");
+    local FlareWait = mob:getLocalVar("FlareWait")
+    local GigaFlare = mob:getLocalVar("GigaFlare");
+    local tauntOneShown = mob:getLocalVar("tauntOneShown");
+    local tauntTwoShown = mob:getLocalVar("tauntTwoShown");
+    local tauntThreeShown = mob:getLocalVar("tauntThreeShown");
+    local mobHPP = mob:getHPP();
+    local isBusy = false;
+    if (GetMobAction(mob:getID()) == 33 or GetMobAction(mob:getID()) == 34 or GetMobAction(mob:getID()) == 11 or GetMobAction(mob:getID()) == 8 or GetMobAction(mob:getID()) == 30 or GetMobAction(mob:getID()) == 4) then
```
Comment (xdemolish) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: use ACTION_ crap here (see status.lua)

File: scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua [hunk 5]
```diff
@@ -1,15 +1,97 @@
...
+                mob:setLocalVar("FlareWait", mob:getBattleTime() + 2); -- second taunt happens two seconds after the first.
+                mob:setLocalVar("tauntOneShown", 1);
+            elseif (FlareWait < mob:getBattleTime() and FlareWait ~= 0 and tauntOneShown == 1) then -- the wait time between the first and second taunt as passed. Checks for wait to be not 0 because it's set to 0 on successful use.
+                if (tauntTwoShown == 0) then
+                    mob:setLocalVar("tauntTwoShown", 1); -- if Megaflare gets stunned it won't show the text again, until successful use.
+                    target:showText(mob,BAHAMUT_TAUNT + 1);
+                end;
+                if (mob:checkDistance(target) <= 15) then -- without this check if the target is out of range it will keep attemping and failing to use Megaflare. Both Megaflare and Gigaflare have range 15.
+                    if (mob:getBehaviour() == 1040) then -- default behaviour
+                        mob:setBehaviour(mob:getBehaviour() - 1024) -- 1024 is noturn
```
Comment (xdemolish) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: https://github.com/DarkstarProject/darkstar/blob/master/scripts/globals/status.lua#L2074

File: scripts/globals/mobskills/Gigaflare.lua [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    return 1;
+end;
+
+function onMobWeaponSkill(target, mob, skill)
+    mob:setLocalVar("GigaFlare", 1); -- When set to 1 the script won't call it.
+    mob:setLocalVar("tauntThreeShown", 0);
+    mob:SetMobAbilityEnabled(true); -- enable the spells/other mobskills again
+    mob:SetMagicCastingEnabled(true);
+    if (mob:getBehaviour() == 16) then -- re-enable noturn
+        mob:setBehaviour(mob:getBehaviour() + 1024) -- 1024 is noturn
```
Comment (xdemolish) [inline scripts/globals/mobskills/Gigaflare.lua]: use the behaviour variable crap here too

**Comment by Hozu** [comment]
@Zenny89 I did the script changes that were requested, I'll send em to you again.

File: sql/mob_pools.sql [hunk 7]
```diff
@@ -5506,7 +5506,7 @@ INSERT INTO `mob_pools` VALUES (5532,'Water_Elemental',106,0x00000C0000000000000
 INSERT INTO `mob_pools` VALUES (5533,'Dark_Elemental',100,0x00000F0000000000000000000000000000000000,8,8,11,240,100,0,32,0,2,65,0,291,131,0,0,18,1,0);
 INSERT INTO `mob_pools` VALUES (5534,'Air_Elemental',99,0x00000A0000000000000000000000000000000000,4,5,12,240,100,0,32,0,2,658,0,291,131,0,0,12,1,0);
 INSERT INTO `mob_pools` VALUES (5535,'Ice_Elemental',103,0x0000090000000000000000000000000000000000,4,5,12,240,100,0,32,0,2,54,0,291,131,0,0,14,1,0);
-INSERT INTO `mob_pools` VALUES (5536,'Aern_s_Wynav',193,0x0000120200000000000000000000000000000000,10,1,7,200,100,0,0,1,8,1,0,0,0,0,1,136,1,0);  -- Ix'DRG's pets
+INSERT INTO `mob_pools` VALUES (5536,'Aern_s_Wynav',193,0x0000120200000000000000000000000000000000,10,1,7,200,100,0,0,1,8,1,0,0,0,0,0,136,1,0);  -- Ix'DRG's pets
```
Comment (TeoTwawki) [inline sql/mob_pools.sql]: Seems unrelated, accidental include?

File: sql/mob_pools.sql [hunk 8]
```diff
@@ -5516,6 +5516,7 @@ INSERT INTO `mob_pools` VALUES (5542,'Eschan_Tarichuk',98,0x00004605000000000000
 INSERT INTO `mob_pools` VALUES (5543,'Eschan_Mosquito',48,0x0000670b00000000000000000000000000000000,1,1,2,240,100,0,257,1,0,0,0,7,131,0,0,0,1,0);
 INSERT INTO `mob_pools` VALUES (5544,'Ghillie_Dhu',437,0x0000880100000000000000000000000000000000,1,1,7,240,110,0,0,1,0,7,32,5449,131,0,0,0,1,0);
 INSERT INTO `mob_pools` VALUES (5545,'Highlander_Lizard',438,0x0000480100000000000000000000000000000000,1,1,7,240,110,0,0,1,0,0,0,980,131,0,0,0,1,0);
+INSERT INTO `mob_pools` VALUES (5546,'Chaneque',496,0x0000C70800000000000000000000000000000000,2,2,7,360,100,0,0,0,2,7,32,0,0,0,0,0,1,0);
```
Comment (TeoTwawki) [inline sql/mob_pools.sql]: Here too.

File: sql/mob_skill.sql (hunk 3 above)
Comment (TeoTwawki) [inline sql/mob_skill.sql]: Ah, yep.

File: sql/mob_pools.sql (hunk 7 above)
Comment (Hozu) [inline sql/mob_pools.sql]: Yeah, must be.

File: sql/mob_skill.sql (hunk 3 above)
Comment (TeoTwawki) [inline sql/mob_skill.sql]: Gotta be careful which branch you grab stuff from, heh.

**Comment by Deadwing888** [comment]
Some things I picked up from my footage (and this is bv2 mind you):

Immune to paralyze
Resistant to slow
Immune to silence
Immune to bind
Immune to gravity
Avg hit to someone in no phys- gear = ~500 dmg
25k Hp
He looks to be a 240 delay.
Fire V, Flare II, and cure V have about a 2 second cast time.
Tried to use sweeping flail when I was 30' away and "no valid targets within area of effect"
Initial aggro range = 20' however if you wipe he has a 45'+ aggro range
I fall party members fall it gives a 3 minute warning till you get kicked out

@Hozu sadly I can not answer either of your questions, I had a buddy with me in the BC but he never stood more than 10' away when an AOE went off. We also never got his hp low enough to use gigaflare. I think I have one last cloud evoker so I might be able to get that info, is there anything else we need to know about this guy?

**Comment by Hozu** [comment]
I wonder how much of that applies to v1. Right now it has sleep/silence/bind immunity. I'd be quite surprised if it hit THAT hard though.

Right now Sweeping Flail is set to 15', should that be increased? How far were you when it started to prepare the move?

I asked about the spell AoE range because I saw a video where Graviga hit more people than I expected. Though maybe I'm just thinking it's larger because it was a view from far away so it could see the AoE range in both directions around the target.

**Comment by Deadwing888** [comment]
bv2
Tried to use when I was 30' away.

sweeping flail
This seems to indicate closer to 20' for sweeping flail.

**Comment by Deadwing888** [comment]
I'm not sure how much that applies to Bv1. If you want my guesses I'd say Bv1 is probably also immune to paralyze grav and sleep and probably does hit for less damage.

If you brought your BV2 alliance to do BV1 it was really a pushover; I did my fight with 18 and it was over in about 4 minutes, I don't think we even bothered killing the wyrm summons.

Bv2 on the other hand was probably the most challenging 18 man instance in classic XI. I remember it being 20 minutes of just slightly controlled chaos. In WopaCo we had our whms sub ninja to blink the aoes and offhand a raise II rod to get people up. The roar also hits everybody, so fulltiming RR equipment was pretty standard.

Talking to a friend of mine he says gigaflare was indeed conal but was a realllly wide cone.

**Comment by Hozu** [comment]
Weird, I guess I'll have to leave Gigaflare as cone, though I don't think there's a way to handle cone angle range on DSP? Also I didn't know that it'd try to use the ability even though it was too far to hit with it. It's currently range 15 so it will have to be bumped up to 20. It'd make sense for it to have paralysis and gravity immunity.

Also Bv1 doesn't summon anything, it's alone. That'd be why you didn't have to deal with anything else. It is a story "mission"/quest with no reward after all.

**Comment by Deadwing888** [comment]
> Also I didn't know that it'd try to use the ability even though it was too far to hit with it. -Hozu

It seemed odd to me too.

> Also Bv1 doesn't summon anything, it's alone. That'd be why you didn't have to deal with anything else. - Hozu

That would explain it.

**Comment by Zenny89** [comment]
@TeoTwawki  No, not accidental include. Those are from a PR the other day and no idea why GitHub is marking them as new changes.

**Comment by TeoTwawki** [comment]
@Zenny89 Actually the logical explanation is: those changes were merged after this PR was opened, this (and your local copy) are based on an earlier revision of DSP master. When you got the changes from Hozu the files he was working with were based of a newer revision that included those. Your local copy has git thinking you added these lines. It hasn't tripped any conflicts because they are 100% identical on those lines.

Shorter version: just rebase it to latest DSP revision and you'll be fine

**Comment by Zenny89** [comment]
What happens if I don't rebase?

**Comment by TeoTwawki** [comment]
It'll continue to look screwy on here, and I have no idea if anything weird happens to the history on merge (I suspect it'll just claim you as last one to touch it on git blame, but no other negative side effects).

**Comment by TeoTwawki** [comment]
Alternatively an amend-commit removing those (if my theory is correct) should be safe too. Rebase was just 1st thing to come to mind.

**Comment by Zenny89** [comment]
Should already be fixed.

**Comment by TeoTwawki** [comment]
Looks like it is.

**Comment by Zenny89** [comment]
Incorporated various comments.

File: scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua [hunk 9]
```diff
@@ -1,15 +1,97 @@
...
+        mob:setLocalVar("MegaFlareTrigger", 7);
+        mob:setLocalVar("MegaFlareQueue", MegaFlareQueue + 1);
+    elseif (mobHPP < 20 and MegaFlareTrigger < 8) then
+        mob:setLocalVar("MegaFlareTrigger", 8);
+        mob:setLocalVar("MegaFlareQueue", MegaFlareQueue + 1);
+    end;
+
+    if (mob:actionQueueEmpty() == true and isBusy == false) then -- the last check prevents multiple Mega/Gigaflares from being called at the same time.
+        if (MegaFlareQueue > 0) then
+            mob:SetMobAbilityEnabled(false); -- disable all other actions until Megaflare is used successfully
```
Comment (Hozu) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: Seems like all three of these have to be set to true in onMobDespawn, since it persists through to the next fight if it was killed while they were disabled.

Comment (TeoTwawki) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: set them true in `onMobDespawn()`, since that runs after death and if a group wipes `onMobDeath()` won't be run...And I am assuming BCNM cleanup won't interfere with that >.>

File: scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua [hunk 10]
```diff
@@ -1,15 +1,97 @@
...
+function onMobFight(mob,target)
+    local MegaFlareQueue = mob:getLocalVar("MegaFlareQueue");
+    local MegaFlareTrigger = mob:getLocalVar("MegaFlareTrigger");
+    local MegaFlareUses = mob:getLocalVar("MegaFlareUses");
+    local FlareWait = mob:getLocalVar("FlareWait")
+    local GigaFlare = mob:getLocalVar("GigaFlare");
+    local tauntShown = mob:getLocalVar("tauntShown");
+    local mobHPP = mob:getHPP();
+    local isBusy = false;
+    if (GetMobAction(mob:getID()) == ACTION_MOBABILITY_START or GetMobAction(mob:getID()) == ACTION_MOBABILITY_USING or GetMobAction(mob:getID()) == ACTION_MOBABILITY_FINISH or GetMobAction(mob:getID()) == ACTION_MAGIC_START or GetMobAction(mob:getID()) == ACTION_MAGIC_CASTING or GetMobAction(mob:getID()) == ACTION_MAGIC_START) then
```
Comment (TeoTwawki) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: I'm just curious: I thought I saw something like `if (mob:actionQueueEmpty() == true) then` in one of the previous revisions to this script. Didn't work here?

Comment (Hozu) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: It didn't seem to be, because many times Bahamut would use Megaflare or Gigaflare twice in a row, with no text the second time.

Comment (TeoTwawki) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: I just realized what a crazy long line that is. HATE right/left scrolling. Would rather see:
`if (GetMobAction(mob:getID()) == ACTION_MOBABILITY_START`
`or GetMobAction(mob:getID()) == ACTION_MOBABILITY_USING`
`or GetMobAction(mob:getID()) == ACTION_MOBABILITY_FINISH`
`or GetMobAction(mob:getID()) == ACTION_MAGIC_START`
`or GetMobAction(mob:getID()) == ACTION_MAGIC_CASTING`
`or GetMobAction(mob:getID()) == ACTION_MAGIC_START) then`
If all that is truly necessary (I'd see if maybe someone could tell me why the other thing didn't work first though).

Comment (teschnei) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: the queue is empty as soon as he starts using the move. it would only be useful here if the conditions for mega/gigaflare were changed when the move was started.  jormungand/ouryu/tiamat use it to make sure they don't touchdown multiple times in a row if the time ticked while casting a spell

Comment (TeoTwawki) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: > the queue is empty as soon as he starts using the move

Oh.

Comment (Hozu) [inline scripts/zones/Riverne-Site_B01/mobs/Bahamut.lua]: And that explains everything.

**Comment by Zenny89** [comment]
What, if anything, needs to be updated on this PR?

**Comment by xdemolish** [comment]
leaving for @teschnei cause AI rewrite

**Comment by AlexandreCA** [comment]
When is it coming out?

**Comment by teschnei** [comment]
closing as I updated it into its own temp branch with merge resolution (to be tested and merged to master after that)
